net/dns/resolver/forwarder.go

```diff
@@ -210,6 +222,57 @@ func (f *forwarder) packetListener(ip netaddr.IP) (packetListener, error) {
 	return lc, nil
 }
 
+func (f *forwarder) getDoHClient(ip netaddr.IP) (urlBase string, c *http.Client, ok bool) {
+	urlBase, ok = knownDoH[ip]
+	if !ok {
+		return
+	}
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if c, ok := f.dohClient[ip]; ok {
+		return urlBase, c, true
+	}
+	if f.dohClient == nil {
+		f.dohClient = map[netaddr.IP]*http.Client{}
+	}
+	nsDialer := netns.NewDialer()
+	c = &http.Client{
+		Transport: &http.Transport{
+			IdleConnTimeout: dohTransportTimeout,
+			DialContext: func(ctx context.Context, netw, addr string) (net.Conn, error) {
+				if !strings.HasPrefix(netw, "tcp") {
+					return nil, fmt.Errorf("unexpected network %q", netw)
+				}
+				return nsDialer.DialContext(ctx, "tcp", net.JoinHostPort(ip.String(), "443"))
+			},
+		},
+	}
+	f.dohClient[ip] = c
+	return urlBase, c, true
+}
+
+const dohType = "application/dns-message"
+
+func (f *forwarder) sendDoH(ctx context.Context, urlBase string, c *http.Client, packet []byte) ([]byte, error) {
+	req, err := http.NewRequestWithContext(ctx, "POST", urlBase, bytes.NewReader(packet))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", dohType)
```

Comment:
I think setting both Accept and Content-Type would be good

Reply:
I see it included in examples (https://datatracker.ietf.org/doc/html/rfc8484#page-5) but looks like the Accept one is pretty optional, though:

> The DoH client SHOULD include an HTTP Accept request header field to
>  indicate what type of content can be understood in response.
>  Irrespective of the value of the Accept request header field, the
> client MUST be prepared to process "application/dns-message" 

(and empirically isn't required by any provider)

Worth the bandwidth? With HTTP/2 it's not much, but iOS doesn't use HTTP/2 yet so it's a bit more.

Reply:
Updated code with comment:

```patch
diff --git a/net/dns/resolver/forwarder.go b/net/dns/resolver/forwarder.go
index 8a15c099..a3184f8c 100644
--- a/net/dns/resolver/forwarder.go
+++ b/net/dns/resolver/forwarder.go
@@ -259,6 +259,12 @@ func (f *forwarder) sendDoH(ctx context.Context, urlBase string, c *http.Client,
                return nil, err
        }
        req.Header.Set("Content-Type", dohType)
+       // Note: we don't currently set the Accept header (which is
+       // only a SHOULD in the spec) as iOS doesn't use HTTP/2 and
+       // we'd rather save a few bytes on outgoing requests when
+       // empirically no provider cares about the Accept header's
+       // absence.
+
        hres, err := c.Do(req)
        if err != nil {
                return nil, err
```